File: android/app/src/main/java/io/openorcha/mobile/domain/DeviceAuthFlow.kt
package io.openorcha.mobile.domain

/**
 * The auth-options sheet's state machine for the GitHub device-token flow -- a pure
 * reducer (no session, no networking) so every transition is testable. Ports
 * `ios/Orcha/Domain/DeviceAuthFlow.swift` 1:1.
 *
 *     Options --SignInTapped--> SigningIn --CallbackReceived--> Connecting --RetrySucceeded--> Connected
 *        ^                        |   |                            |
 *        +-------Cancelled--------+   +-InvalidCallback-> Failed <-+--RetryFailed
 *
 * `Failed` renders as the options sheet with a danger banner; `SignInTapped` from
 * there starts the round-trip over. A user cancel goes back to plain `Options` with
 * no banner -- closing the Custom Tab is their own action (and a server-side mint
 * failure ends the session the same way).
 */
data class DeviceAuthFlow(val phase: Phase = Phase.Options) {

    sealed interface Phase {
        /** Showing the options: Sign in with GitHub, or the collapsed token entry. */
        data object Options : Phase

        /** The Custom Tab is up; waiting for the callback (or a close). */
        data object SigningIn : Phase

        /** Callback parsed -- retrying the captured pairing probe with the minted token. */
        data object Connecting : Phase

        /** Back on the options with a banner explaining what went wrong. */
        data class Failed(val message: String) : Phase

        /** The retry connected; the sheet dismisses. */
        data object Connected : Phase
    }

    sealed interface Event {
        data object SignInTapped : Event

        /**
         * The Custom Tab returned an `orcha://auth/callback` URI that parsed and
         * named the pending deployment.
         */
        data object CallbackReceived : Event

        /**
         * The user closed the Custom Tab -- or the server showed its own error
         * page and never redirected, ending the session the same way.
         */
        data object Cancelled : Event

        /** The session returned a URI that didn't parse or named another host. */
        data object InvalidCallback : Event

        data object RetrySucceeded : Event

        /**
         * The attempt failed with a reason worth a banner: the minted token was
         * rejected on the probe retry, or the flow couldn't even start.
         */
        data class RetryFailed(val message: String) : Event
    }

    /**
     * Advance the machine. Events that make no sense in the current phase are
     * ignored (e.g. a stray `Cancelled` after the retry already began).
     */
    fun handle(event: Event): DeviceAuthFlow {
        val next = when (phase) {
            is Phase.Options -> when (event) {
                is Event.SignInTapped -> Phase.SigningIn
                else -> null
            }
            is Phase.SigningIn -> when (event) {
                is Event.Cancelled -> Phase.Options
                is Event.CallbackReceived -> Phase.Connecting
                is Event.InvalidCallback -> Phase.Failed(
                    "GitHub finished, but the sign-in didn't come back as expected. Try again, or use an access token instead.",
                )
                is Event.RetryFailed -> Phase.Failed(event.message)
                else -> null
            }
            is Phase.Connecting -> when (event) {
                is Event.RetrySucceeded -> Phase.Connected
                is Event.RetryFailed -> Phase.Failed(event.message)
                else -> null
            }
            is Phase.Failed -> when (event) {
                is Event.SignInTapped -> Phase.SigningIn
                else -> null
            }
            is Phase.Connected -> null
        }
        return if (next == null) this else copy(phase = next)
    }
}

File: ios/Orcha/Domain/DeviceAuthFlow.swift
import Foundation

/// The auth-options sheet's state machine for the GitHub device-token flow —
/// a pure reducer (no session, no networking) so every transition is testable.
///
///     options ──signInTapped──▶ signingIn ──callbackReceived──▶ connecting ──retrySucceeded──▶ connected
///        ▲                        │   │                            │
///        └───────cancelled────────┘   └─invalidCallback─▶ failed ◀─┴─retryFailed
///
/// `failed` renders as the options sheet with a danger banner; `signInTapped`
/// from there starts the round-trip over. A user cancel goes back to plain
/// `options` with no banner — closing the browser sheet is their own action
/// (and a server-side mint failure ends the session the same way).
struct DeviceAuthFlow: Equatable {
    enum Phase: Equatable {
        /// Showing the options: Sign in with GitHub, or the collapsed token entry.
        case options
        /// The browser session is up; waiting for the callback (or a close).
        case signingIn
        /// Callback parsed — retrying the captured pairing probe with the minted token.
        case connecting
        /// Back on the options with a banner explaining what went wrong.
        case failed(String)
        /// The retry connected; the sheet dismisses.
        case connected
    }

    enum Event: Equatable {
        case signInTapped
        /// The session returned an `orcha://auth/callback` URL that parsed
        /// and named the pending deployment.
        case callbackReceived
        /// The user closed the browser sheet — or the server showed its own
        /// error page and never redirected, ending the session the same way.
        case cancelled
        /// The session returned a URL that didn't parse or named another host.
        case invalidCallback
        case retrySucceeded
        /// The attempt failed with a reason worth a banner: the minted token
        /// was rejected on the probe retry, or the flow couldn't even start.
        case retryFailed(String)
    }

    private(set) var phase: Phase = .options

    /// Advance the machine. Events that make no sense in the current phase
    /// are ignored (e.g. a stray `cancelled` after the retry already began).
    mutating func handle(_ event: Event) {
        switch (phase, event) {
        case (.options, .signInTapped), (.failed, .signInTapped):
            phase = .signingIn
        case (.signingIn, .cancelled):
            phase = .options
        case (.signingIn, .callbackReceived):
            phase = .connecting
        case (.signingIn, .invalidCallback):
            phase = .failed("GitHub finished, but the sign-in didn't come back as expected. Try again, or use an access token instead.")
        case (.signingIn, .retryFailed(let message)), (.connecting, .retryFailed(let message)):
            phase = .failed(message)
        case (.connecting, .retrySucceeded):
            phase = .connected
        default:
            break
        }
    }
}
